File: TP03/Q6 Pilha Flexivel C/PilhaFlexivel.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <err.h>

typedef struct 
{
   char nome[20];
   int altura;
   double peso;
   char cor_do_cabelo[50];
   char cor_da_pele[50];
   char cor_dos_olhos[50];
   char ano_aniversario[50];
   char genero[10];
   char home_world[50];
}Personagem;

typedef struct Celula
{
   Personagem elemento;
   struct Celula* prox;
}Celula;


// PROTOTIPOS
Personagem LerArquivo (Personagem P, char entrada[]);
Personagem SepararPalavras (Personagem P, char leitura_arq[]);
char* TiraVirgula (char s[]);
Celula* novaCelula (Personagem pers);
void startPers ();
void InserirFim (Personagem novo);
void RemoverFim ();
void Imprimir (Celula* i, int j);
void ImprimirRemovidos (Celula* i);
void Remover ();
void SepararComando (char entrada[], Personagem novo);
void StartR();
void ImprimirRec();
void ImprimirRemovidosRec();


// VARIAVEIS GLOBAIS
int n = 0; // quantidade de personagens
int qnt_removidos = 0;
Celula* topoP;
Celula* topoR;

int main()
{
   Personagem P;
   //char* entrada = (char*)malloc(sizeof(char)*100);
   char entrada[1000];
   startPers();
   StartR();
 
   fgets (entrada, 100, stdin);
   entrada[strlen(entrada) - 1] = '\0';
   
   while (!strcmp (entrada, "FIM") == 0)
   {  
      P = LerArquivo (P, entrada);
      InserirFim(P); 
      //printf ("\ntopo: %s", topoP->elemento.nome);
      fgets (entrada, 100, stdin);
      entrada[strlen(entrada) - 1] = '\0';
   }

   int qnt = 0;
   int cont = 0;
   scanf ("%d\n", &qnt);
   fgets (entrada, 100, stdin);
   entrada[strlen(entrada) - 1] = '\0';
   char aux[200];
   
   //printf ("\ntopo: %s", topoP->elemento.nome);

   while (cont < qnt)
   {
      if (strstr(entrada, " ") != NULL)
      {
         SepararComando(entrada, P);
      }
      else
      {
         RemoverFim();
      }

      fgets (entrada, 100, stdin);
      entrada[strlen(entrada) - 1] = '\0';
      cont++;
   }

   ImprimirRemovidosRec ();
   ImprimirRec ();
   return 0;
}


/*
 * Essa função vai realizar a leitura do arquivo enviado
 */
Personagem LerArquivo (Personagem P, char entrada[])
{
   char leitura_arq[1000];
   FILE *ptr_arq;
   ptr_arq = fopen(entrada, "r");

   while (!feof(ptr_arq))
   {
      fscanf (ptr_arq, "%[^\n]\n", leitura_arq);
   }
   fclose (ptr_arq);

   return SepararPalavras(P, leitura_arq);
}


/*
 * Essa função vai separar os Atributos do personagem
 */
Personagem SepararPalavras (Personagem P, char leitura_arq[])
{  
   char *pers = leitura_arq;
   int k = 0;
   char string[50][20];
   char *ptr_separado;
   const char separar[2] = "'";
   
   ptr_separado = strtok (pers, separar);
   
   //while (ptr_separado != NULL)
   for (int i = 0; i < 36; i++)
   {
      strcpy(string[k], ptr_separado);
      ptr_separado = strtok(NULL, separar);
      k++;
   }
   
   for (int i = 0; i < 36; i++)
   {
      if (strcmp (string[i], "name") == 0)
      {
         strcpy (P.nome, string[i+2]);
      }
      else if (strcmp (string[i], "height") == 0)
      {
         P.altura = atoi(string[i+2]);
      }
      else if (strcmp (string[i], "mass") == 0)
      {
         P.peso = atof(TiraVirgula(string[i+2]));
      }
      else if (strcmp (string[i], "hair_color") == 0)
      {
         strcpy (P.cor_do_cabelo, string[i+2]);
      }
      else if (strcmp(string[i], "skin_color") == 0)
		{
			strcpy (P.cor_da_pele, string[i+2]);
		}
		else if (strcmp(string[i], "eye_color") == 0)
		{
			strcpy (P.cor_dos_olhos, string[i+2]);
		}
		else if (strcmp(string[i], "birth_year") == 0)
		{
			strcpy (P.ano_aniversario, string[i+2]);
		}
		else if (strcmp(string[i], "gender") == 0)
		{
			strcpy (P.genero, string[i+2]);
		}
		else if (strcmp(string[i], "homeworld") == 0)
		{
			strcpy (P.home_world, string[i+2]);
		}
   }
   return P;
}


/*
 * Essa função recebe uma string e remove dela as virgulas
 */
char* TiraVirgula (char s[])
{
   char *aux = (char*)malloc(10*sizeof(char));
   int cont = 0;
   for (int i = 0; i < strlen(s); i++)
   {
      if (!(s[i] == ','))
      {
         aux[cont] = s[i];
         cont++;
      }
   }
   return aux;
}


/*
 * Essa função vai separar o comando de inserir e o arquivo
 */
void SepararComando (char entrada[], Personagem novo)
{
   char *ptr_entrada = entrada;
   char *ptr_separado;
   const char separar[2] = " ";
   char string[10][20];
   int k = 0;
   
   ptr_separado = strtok(ptr_entrada, separar);  
   
   while (ptr_separado != NULL)
   {
      strcpy(string[k], ptr_separado);
      ptr_separado = strtok(NULL, separar); 
      k++;
   }
   novo = LerArquivo(novo, string[1]);
   InserirFim(novo);

   
}

/*
 * Essa função vai criar minha celula
 */
Celula* novaCelula (Personagem elemento)
{
   Celula* nova = (Celula*)malloc(sizeof(Celula));
   nova->elemento = elemento;
   nova->prox = NULL;
   return nova;
}

/*
 * Essa função vai iniciar minha lista vazia
 */
void startPers()
{
   topoP = NULL;
}
void StartR()
{
   topoR = NULL;
}

 /*
  * Essa função vai inserir um personagem no final da lista
  */
void InserirFim (Personagem novo)
{
   Celula* temp = novaCelula(novo);
   temp->prox = topoP;
   topoP = temp;
   temp = NULL;
   n++;
   //printf ("\ntopo: %s", topoP->elemento.nome);
}


/*
 * Essa função vai remover o personagem do fim da lista
 */
void RemoverFim ()
{
    //printf ("\ntopo antes da remoção: %s", topoP->elemento.nome);
   
    if (topoP == NULL)
   {
      printf ("erro");
   }
   else
   {
      // inserir na minha pilha de removidos
      Celula* temp = topoP;
      topoP = topoP->prox;
      temp->prox = topoR;
      topoR = temp;
      temp = NULL;
      n--;
      //printf ("\ntopo depois da remoção: %s", topoP->elemento.nome);
      //printf ("\ntopoRemovidos depois da remoção: %s\n", topoR->elemento.nome);

   }
}

void Remover()
{
  printf ("\ntopo: %s", topoP->elemento.nome);
 

}
/*
 * Essa função vai imprimir os atributos dos personagens
 */
void ImprimirRec()
{
   Celula* i = topoP;
   Imprimir(i, n);
}

void Imprimir (Celula* i, int j) 
{
  
   if(i != NULL)
   {
      Imprimir(i->prox, --j);
      printf ("[%i]  ## %s ## ", j, i->elemento.nome);
		printf ("%i ## ", i->elemento.altura);
		printf ("%g ## ", i->elemento.peso);
		printf ("%s ## ", i->elemento.cor_do_cabelo);
		printf ("%s ## ", i->elemento.cor_da_pele);
		printf ("%s ## ", i->elemento.cor_dos_olhos);
		printf ("%s ## ", i->elemento.ano_aniversario);
		printf ("%s ## ", i->elemento.genero);
		printf ("%s ## ", i->elemento.home_world);
      printf ("\n");	
   }
}


/*
 * Essa função vai imprimir o nome dos personagens removidos 
 */
void ImprimirRemovidosRec()
{
   Celula* i = topoR;
   ImprimirRemovidos(i);
}

void ImprimirRemovidos (Celula* i)
{
   if (i != NULL)
   {
      ImprimirRemovidos(i->prox);
      printf ("(R) %s\n", i->elemento.nome);
   }
   
}

File: TP03/Q2 Pilha felixivel java/PilhaFlexivel.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

class PilhaFlexivel
{
   public static Pilha pilha = new Pilha();
   
   public static void main (String[] args) throws Exception
   {
      MyIO.setCharset("UTF-8"); 
      String entrada = "";
      entrada = MyIO.readLine();

      while(!entrada.contains("FIM"))
      {  
         Personagem personagem = new Personagem();
         personagem.LeArquivo(entrada);
         pilha.Empilhar(personagem);
         entrada = MyIO.readLine();
      }
        
      int modificacoes = MyIO.readInt();
      Processos(modificacoes);      
      pilha.mostrarRe();
      pilha.MostrarPers();
   }  
    
   /*
    * Esse metodo vai analisar qual processo da lista vai ser executado
    */
   public static void Processos (int modificacoes) throws Exception 
   {
      String in = "";
      String[] aux = new String[500];
       
      for (int i = 0; i < modificacoes; i++)
      {
         in = MyIO.readLine();
               
         if (in.contains(" "))
         {                    
            aux = in.split(" ");
            Personagem novo = new Personagem();
            novo.LeArquivo(aux[1]);
            pilha.Empilhar(novo);
         }else
         {
            aux[0] = in;
            pilha.Desempilhar();
         }
      }
   }
}
class Personagem
{
    // ATRIBUTOS 
    private String frase_completa = "";
    private String nome = "";
    private int altura = 0;
    private double peso = 0;
    private String corDoCabelo = "";
    private String corDaPele = "";
    private String corDosOlhos = "";
    private String anoNascimento = "";
    private String genero = "";
    private String homeWorld = "";
    public Personagem prox;

 
    // CONSTRUTOR
    Personagem(String nome, int altura, double peso, String corDoCabelo, String corDaPele, String corDosOlhos, String anoNascimento, String genero, String homeWorld){  
       this.frase_completa = frase_completa;
       this.nome = nome;
       this.altura = altura;
       this.peso = peso;
       this.corDoCabelo = corDoCabelo;
       this.corDaPele = corDaPele;
       this.corDosOlhos = corDosOlhos;
       this.anoNascimento = anoNascimento;
       this.genero = genero;
       this.homeWorld = homeWorld;
       this.prox = null;
    } 
     
    // CONSTRUTOR VAZIO
    Personagem ()
    {
    }
       
    // METODOS

// -------------------------------------------SET'S AND GET'S----------------------------------------------------------------------------- 
    public void setfrase_completa (String frase_completa)
    {
       this.frase_completa = frase_completa;
    }
    public void setnome (String nome)
    {
       this.nome = nome;
    }
    public void setaltura (int altura)
    {
       this.altura = altura;
    }
    public void setpeso (double peso)
    {
       this.peso = peso;
    }
    public void setcorDoCabelo (String corDoCabelo)
    {
       this.corDoCabelo = corDoCabelo;
    }
    public void setcorDaPele (String corDaPele)
    {
       this.corDaPele = corDaPele;
    }
    public void setcorDosOlhos (String corDosOlhos)
    {
       this.corDosOlhos = corDosOlhos;
    }
    public void setanoNascimento (String anoNascimento)
    {
       this.anoNascimento = anoNascimento;
    }
    public void setgenero (String genero)
    {
       this.genero = genero;
    }
    public void sethomeWorld (String homeWorld)
    {
       this.homeWorld = homeWorld;
    }
    public String getfrase_completa()
    {
       return this.frase_completa;
    }
    public String getnome ()
    {
       return this.nome;
    }
    public int getaltura()
    {
       return this.altura;
    }
    public double getpeso()
    {
       return this.peso;
    }
    public String getcorDoCabelo()
    {
       return this.corDoCabelo;
    }
    public String getcorDaPele()
    {
       return this.corDaPele;
    }
    public String getcorDosOlhos()
    {
       return this.corDosOlhos;
    }
    public String getanoNascimeto()
    {
       return this.anoNascimento;
    }
    public String getgenero()
    {
       return this.genero;
    }
    public String gethomeWorld()
    {
       return this.homeWorld;
    }
//-----------------------------------i-------------END OF SET'S E GET'S-------------------------------------------------------------------
    
    // METODO PARA LEITUA DO ARQUIVO TXT
    
    public void LeArquivo (String patch)
    {
       String dados_do_personagem = "";

       try
       {
          FileReader arq = new FileReader(patch); // abrindo o arquivo 
          BufferedReader ler_arq = new BufferedReader(arq); // lendo meu arquivo

          String o_que_ta_dentro_do_arquivo = ler_arq.readLine();
          
          while (ler_arq.ready())
          {
             o_que_ta_dentro_do_arquivo = ler_arq.readLine(); // é uma string só, nao precisa concatenar
             
          }

          this.frase_completa = o_que_ta_dentro_do_arquivo;
          arq.close();
          ler_arq.close();
          
       }catch (IOException ex)
       {
          System.err.printf("Erro %s.\n", ex.getMessage());
       }

       SepararDados();

    } // FIM DA LEITURA DO ARQUIVO

    // METODO PARA SEPARAR OS DADOS DO PERSONAGEM

    public void SepararDados()
    {
       String[] frase_separada = this.frase_completa.split("',"); // A frase separada pega por exemplo "}'name': 'Ackabar'
       String[] caracteristicas = new String[2000]; // a caracteristica vai pegar apenas o nome ou altura...
       
       for (int i = 0; i <= 8; i++) // usei menor ou igual a oito porque cada entrada tem, que me interessa,
       {                            //9 strings separadas por 8 virgulas
                    
          caracteristicas = frase_separada[i].split ("'");

          //OBS: switch (caracteristicas[i]) SWITCH NAO TRABALHA COM BOOLEAN
          
          if (caracteristicas[1].contains("name"))
          {
                setnome(caracteristicas[3]);
          }
          else if (caracteristicas[1].contains("height"))
          {
                if (caracteristicas[3].contains("unknow"))
                {
                   setaltura(0);
                }
                else setaltura(Integer.parseInt(caracteristicas[3]));
                
          }
          else if (caracteristicas[1].contains("mass"))
          {
                if (caracteristicas[3].contains("unknow"))
                {
                   setpeso(0);
                }
                else setpeso(Double.parseDouble(TiraVirgula(caracteristicas[3])));
          }
          else if (caracteristicas[1].contains("hair_color"))
          {
                setcorDoCabelo(caracteristicas[3]);
          }
          else if (caracteristicas[1].contains("skin_color"))
          {
                setcorDaPele(caracteristicas[3]);
          }
          else if (caracteristicas[1].contains("eye_color"))
          {
                setcorDosOlhos(caracteristicas[3]);
          }
          else if (caracteristicas[1].contains("birth_year"))
          {
                setanoNascimento(caracteristicas[3]);
          }
          else if (caracteristicas[1].contains("gender"))
          {
                setgenero(caracteristicas[3]);
          }
          else if (caracteristicas[1].contains("homeworld"))
          {
                sethomeWorld (caracteristicas[3]);
          }


       }// FIM DO FOR
    }// FIM DO METODO SEPARAR DADOS

    public static String TiraVirgula(String s)
    {
        String aux = "";
        for (int i = 0; i < s.length(); i++)
        {
            if (s.charAt(i) != ',')
            {
                 aux += s.charAt(i);
            }
        }
        return aux;
    }

    public Personagem Clone()
    {  
        Personagem p = new Personagem();
    
        p.setnome(this.nome);
        p.setaltura(this.altura);
        p.setpeso(this.peso);
        p.setcorDoCabelo(this.corDoCabelo);
        p.setcorDaPele(this.corDaPele);
        p.setcorDosOlhos(this.corDosOlhos);
        p.setanoNascimento(this.anoNascimento);
        p.setgenero(this.genero);
        p.sethomeWorld(this.homeWorld);

        return p;
    }

    public void imprimir() 
    {
       MyIO.println (" ## " + this.nome + " ## " + this.altura + " ## "  + String.valueOf(this.peso).replace(".0","") + " ## " + this.corDoCabelo + " ## " + this.corDaPele         + " ## " + this.corDosOlhos + " ## " + this.anoNascimento + " ## " + this.genero + " ## " + this.homeWorld + " ## ");             
    }
 

} // FIM DA CLASSE PERSONAGENS

class Celula
{
    public Personagem elemento;
    public Celula prox;

    public Celula()
    {
        this (null);
    }

    public Celula (Personagem elemento)
    {
        this.elemento = elemento;
        this.prox = null;
    }
}
class Pilha
{
   private Celula topo;
   private Celula topoRe; // celula que guarda o topo da pilha dos personagens removidos
    
   public Pilha() 
   {
      topo = new Celula();
      topoRe = new Celula();
   }

        
   /*
    * Esse metodo vai inserir um personagem na minha pilha
    */
   public void Empilhar (Personagem novo) 
   {
      Celula temp = new Celula(novo);
      temp.prox = topo;
      topo = temp;
      temp = null;
   }
   
   /*
    * Esse metodo vai remover um personagem do final da lista
    */
   public void Desempilhar() throws Exception 
   { 
      Celula i = topo;
      if (topo == null) 
      {
			throw new Exception("Erro ao remover!");
		}
      
      topo = topo.prox;
      i.prox = null;

      // Incluir a celula i na pilha de removidos

      i.prox = topoRe;
      topoRe = i;
      i = null;
   } 

   /*
    * Esse metodo vai mostrar os elementos da lista
    */
   public void mostrarRe()
   {
      Celula i = topoRe;
      mostrarRemovidos(i);
   }
   
   public void mostrarRemovidos (Celula i)
   {
      if (i.prox != null)
      {
         mostrarRemovidos(i.prox);
         MyIO.print ("(R) ");
         MyIO.println (i.elemento.getnome());
      }
   }

   public void MostrarPers()
   {
      int tamanho = tamanho();
      Celula i = topo;
      mostrar(i, tamanho);
   }
   public void mostrar(Celula i, int tamanho)
   {
      if (i.prox != null)
      {
         mostrar(i.prox, --tamanho);
         MyIO.print ("[" + tamanho + "]");
         i.elemento.imprimir();
      }
   }

   /*
    * Esse metodo vai verificar o tamanho da lista
    */
   public int tamanho()
   {
      int tamanho = 0;

      for (Celula i = topo; i.prox != null; i = i.prox, tamanho++);
      return tamanho;
   }
}
